test/NuGet.Core.Tests/NuGet.Protocol.Tests/LocalResourceTests/LocalPackageSearchResourceTests.cs

```diff
@@ -586,18 +587,53 @@ public async Task LocalPackageSearchResource_FileSourceAsync()
         }
 
         [Fact]
-        public async Task LocalPackageSearch_SearchAsync_WithCancellationToken_ThrowsAsync()
+        public async Task LocalPackageSearch_SearchAsync_WithCancellationToken_ImmediatelyThrowsAsync()
         {
             using (var root = TestDirectory.Create())
             {
+                // Arrange
                 var localResource = new FindLocalPackagesResourceV2(root);
                 LocalPackageSearchResource resource = new LocalPackageSearchResource(localResource);
 
-                await Assert.ThrowsAsync<OperationCanceledException>(
+                // Act & Assert
+                await Assert.ThrowsAsync<TaskCanceledException>(
                     async () => await resource.SearchAsync("", null, 0, 1, NullLogger.Instance, new CancellationToken(canceled: true)));
             }
         }
 
+        [Fact]
+        public async Task LocalPackageSearch_SearchAsync_SlowLocalRepository_WithCancellationToken_ThrowsAsync()
+        {
+            using (var pathContext = new SimpleTestPathContext())
+            {
+                // Arrange
+                using CancellationTokenSource cts = new();
+                var localRepository = new Mock<FindLocalPackagesResourceV2>(pathContext.PackageSource);
+                localRepository.Setup(f => f.GetPackages(It.IsAny<ILogger>(), It.IsAny<CancellationToken>()))
+                       .Callback((ILogger _, CancellationToken cancellationToken) =>
+                       {
+                           Thread.Sleep(5000);
+                           cancellationToken.ThrowIfCancellationRequested();
+                       }
+                       ).Returns(Enumerable.Empty<LocalPackageInfo>);
+                LocalPackageSearchResource resource = new LocalPackageSearchResource(localRepository.Object);
+
+                // Act
+                Task delay = Task.Delay(TimeSpan.FromMilliseconds(500), cts.Token);
+                Task searchTask = resource.SearchAsync("", null, 0, 1, NullLogger.Instance, cts.Token);
+
+                // Assert
+                var completed = await Task.WhenAny(searchTask, delay);
+                if (completed != delay)
```

Comment:
I see this is because 5000 > 500 milliseconds. Can you add a comment? It's not obvious why "completed != delay" warrants a timeout. 
Alternatively, a message in the exception to explain the situation would be good. Otherwise, if I saw this in a test failure, I'd think we had some bug with timeout logic.

Reply:
Sure, I added comments. Please check if it's clear for someone reading code later.